{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from src.sup_data_to_fasta import load_xlsx\n",
    "\n",
    "sup5_file_path = Path(\"data/Liu_sup5_data.xlsx\")\n",
    "alignments_files_paths = [\n",
    "    Path(\"output/EP-OD0_alignments_results.tsv\"),\n",
    "    Path(\"output/ESP-OD2_alignments_results.tsv\"),\n",
    "    Path(\"output/SP-OD2_alignments_results.tsv\")\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Análise Exploratória dos Dados Suplementares 5\n",
    "\n",
    "## Sobre\n",
    "\n",
    "Este notebook contém uma análise exploratória dos dados fornecidos por Liu, et al. No [suplemento 5](https://www.nature.com/articles/s41467-023-43632-1#additional-information\n",
    ") com o objetivo de averiguar os tratamentos executados no artigo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_scenario_column(df_dict):\n",
    "    return {k:df.assign(scenario=k) for k,df in df_dict.items()}\n",
    "\n",
    "def extract_scenario_from_file_path(string_arg):\n",
    "    return re.search(r\"(.*)-(.*)\", string_arg).group(1)\n",
    "\n",
    "def concat_df_dict(df_dict):\n",
    "    return pd.concat(df_dict.values())\n",
    "\n",
    "alignments_df_list = [pd.read_csv(file_path, sep='\\t') for file_path in alignments_files_paths]\n",
    "alignments_df_dict = {extract_scenario_from_file_path(k.name):v for k,v in zip(alignments_files_paths, alignments_df_list)}\n",
    "alignments_df_dict = add_scenario_column(alignments_df_dict)\n",
    "alignments_df = concat_df_dict(alignments_df_dict)\n",
    "\n",
    "sup5_data_df_dict = load_xlsx(sup5_file_path)\n",
    "sup5_data_df_dict = {extract_scenario_from_file_path(k):v for k,v in sup5_data_df_dict.items()}\n",
    "sup5_data_df_dict = add_scenario_column(sup5_data_df_dict)\n",
    "sup5_data_df = concat_df_dict(sup5_data_df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>qgi</th>\n",
       "      <th>qacc</th>\n",
       "      <th>qaccver</th>\n",
       "      <th>qlen</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>sallseqid</th>\n",
       "      <th>sgi</th>\n",
       "      <th>sallgi</th>\n",
       "      <th>sacc</th>\n",
       "      <th>...</th>\n",
       "      <th>sscinames</th>\n",
       "      <th>scomnames</th>\n",
       "      <th>sblastnames</th>\n",
       "      <th>sskingdoms</th>\n",
       "      <th>stitle</th>\n",
       "      <th>salltitles</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qcovhsp</th>\n",
       "      <th>scenario</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>0</td>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>190</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>0</td>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>rpl19(SL1344_2646)</td>\n",
       "      <td>189</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STnc2180(ncRNA0291)</td>\n",
       "      <td>0</td>\n",
       "      <td>STnc2180(ncRNA0291)</td>\n",
       "      <td>STnc2180(ncRNA0291)</td>\n",
       "      <td>44</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SdsR(ncRNA0070)</td>\n",
       "      <td>0</td>\n",
       "      <td>SdsR(ncRNA0070)</td>\n",
       "      <td>SdsR(ncRNA0070)</td>\n",
       "      <td>78</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>minus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>0</td>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>59</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>NC_003197.2 Salmonella enterica subsp. enteric...</td>\n",
       "      <td>plus</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                qseqid  qgi                 qacc              qaccver  qlen  \\\n",
       "0   rpl19(SL1344_2646)    0   rpl19(SL1344_2646)   rpl19(SL1344_2646)   190   \n",
       "1   rpl19(SL1344_2646)    0   rpl19(SL1344_2646)   rpl19(SL1344_2646)   189   \n",
       "2  STnc2180(ncRNA0291)    0  STnc2180(ncRNA0291)  STnc2180(ncRNA0291)    44   \n",
       "3      SdsR(ncRNA0070)    0      SdsR(ncRNA0070)      SdsR(ncRNA0070)    78   \n",
       "4      ArcZ(ncRNA0002)    0      ArcZ(ncRNA0002)      ArcZ(ncRNA0002)    59   \n",
       "\n",
       "        sseqid    sallseqid  sgi  sallgi         sacc  ... sscinames  \\\n",
       "0  NC_003197.2  NC_003197.2    0       0  NC_003197.2  ...       NaN   \n",
       "1  NC_003197.2  NC_003197.2    0       0  NC_003197.2  ...       NaN   \n",
       "2  NC_003197.2  NC_003197.2    0       0  NC_003197.2  ...       NaN   \n",
       "3  NC_003197.2  NC_003197.2    0       0  NC_003197.2  ...       NaN   \n",
       "4  NC_003197.2  NC_003197.2    0       0  NC_003197.2  ...       NaN   \n",
       "\n",
       "  scomnames  sblastnames  sskingdoms  \\\n",
       "0       NaN          NaN         NaN   \n",
       "1       NaN          NaN         NaN   \n",
       "2       NaN          NaN         NaN   \n",
       "3       NaN          NaN         NaN   \n",
       "4       NaN          NaN         NaN   \n",
       "\n",
       "                                              stitle  \\\n",
       "0  NC_003197.2 Salmonella enterica subsp. enteric...   \n",
       "1  NC_003197.2 Salmonella enterica subsp. enteric...   \n",
       "2  NC_003197.2 Salmonella enterica subsp. enteric...   \n",
       "3  NC_003197.2 Salmonella enterica subsp. enteric...   \n",
       "4  NC_003197.2 Salmonella enterica subsp. enteric...   \n",
       "\n",
       "                                          salltitles  sstrand qcovs qcovhsp  \\\n",
       "0  NC_003197.2 Salmonella enterica subsp. enteric...    minus   100     100   \n",
       "1  NC_003197.2 Salmonella enterica subsp. enteric...    minus   100     100   \n",
       "2  NC_003197.2 Salmonella enterica subsp. enteric...    minus   100     100   \n",
       "3  NC_003197.2 Salmonella enterica subsp. enteric...    minus   100     100   \n",
       "4  NC_003197.2 Salmonella enterica subsp. enteric...     plus   100     100   \n",
       "\n",
       "   scenario  \n",
       "0        EP  \n",
       "1        EP  \n",
       "2        EP  \n",
       "3        EP  \n",
       "4        EP  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def should_be_only_one_genome(df):\n",
    "    return df.sseqid.nunique() == 1\n",
    "\n",
    "def query_start_should_always_be_less_than_query_end(df):\n",
    "    return (df.qstart <= df.qend).all()\n",
    "\n",
    "# validationg\n",
    "assert should_be_only_one_genome(alignments_df), \"o banco de dados deveria ter apenas um genoma de referência de Salmonella\" # all alignments are done to same Salmonella enterica genome\n",
    "assert query_start_should_always_be_less_than_query_end(alignments_df), \"O ínicio da query deve ser menor ou igual ao fim da query\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'Query_ID'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_118566/2593655314.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mdef\u001b[0m \u001b[0mnot_aligned\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignments_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mQuery_ID\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mRNA1_not_aligned\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnot_aligned\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msup5_data_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"RNA1 name\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_118566/2593655314.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(col)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mnot_aligned\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignments_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mQuery_ID\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6200\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6201\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'Query_ID'"
     ]
    }
   ],
   "source": [
    "def not_aligned(col):\n",
    "    return col[~col.isin(alignments_df.Query_ID)]\n",
    "\n",
    "RNA1_not_aligned = not_aligned(sup5_data_df[\"RNA1 name\"])\n",
    "RNA2_not_aligned = not_aligned(sup5_data_df[\"RNA2 name\"])\n",
    "\n",
    "filtered_alignments_df = alignments_df[alignments_df['qseqid'].isin(RNA1_not_aligned) & alignments_df['Query_ID'].isin(RNA2_not_aligned)]\n",
    "print(f\"# queries não alinhadas: {len(filtered_alignments_df)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from pathlib import Path\n",
    "import gzip\n",
    "\n",
    "# Define the path to the genome file\n",
    "genome_file_path = Path(\"data/salmonella_genome.fna\")\n",
    "\n",
    "# Open the gzipped file and parse the sequences\n",
    "with open(genome_file_path, \"rt\") as file:\n",
    "    sequences = SeqIO.parse(file, \"fasta\")\n",
    "    seqs = list(sequences)\n",
    "    salmonella_genome = seqs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_118566/308561010.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  alignments_df[\"Subject_Seq\"] = alignments_df.get([\"Subject_Start\", \"Subject_End\"]).apply(lambda r: salmonella_genome.seq[r[0]-1:r[1]], axis=1)\n"
     ]
    }
   ],
   "source": [
    "# use Subject_Start and End to retrieve the sequence from the genome\n",
    "\n",
    "alignments_df[\"Subject_Seq\"] = alignments_df.get([\"Subject_Start\", \"Subject_End\"]).apply(lambda r: salmonella_genome.seq[r[0]-1:r[1]], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4857450"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(salmonella_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1248179 - 1248137"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query_ID</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>PIdentity</th>\n",
       "      <th>Alignment_Length</th>\n",
       "      <th>Mismatches</th>\n",
       "      <th>Gap_Openings</th>\n",
       "      <th>Query_Start</th>\n",
       "      <th>Query_End</th>\n",
       "      <th>Subject_Start</th>\n",
       "      <th>Subject_End</th>\n",
       "      <th>E_value</th>\n",
       "      <th>Bit_Score</th>\n",
       "      <th>scenario</th>\n",
       "      <th>Subject_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>3490451</td>\n",
       "      <td>3490508</td>\n",
       "      <td>5.090000e-25</td>\n",
       "      <td>108.0</td>\n",
       "      <td>EP</td>\n",
       "      <td>(A, T, T, T, C, C, C, T, G, G, T, G, T, T, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STnc2110(ncRNA0286)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>4009394</td>\n",
       "      <td>4009346</td>\n",
       "      <td>3.940000e-20</td>\n",
       "      <td>91.6</td>\n",
       "      <td>EP</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CpxQ(ncRNA0205)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>4271116</td>\n",
       "      <td>4271176</td>\n",
       "      <td>1.180000e-26</td>\n",
       "      <td>113.0</td>\n",
       "      <td>EP</td>\n",
       "      <td>(G, T, T, T, T, C, C, T, T, G, C, C, A, T, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STnc2000(ncRNA0278)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "      <td>1248179</td>\n",
       "      <td>1248137</td>\n",
       "      <td>7.110000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>EP</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>3490451</td>\n",
       "      <td>3490508</td>\n",
       "      <td>5.090000e-25</td>\n",
       "      <td>108.0</td>\n",
       "      <td>EP</td>\n",
       "      <td>(A, T, T, T, C, C, C, T, G, G, T, G, T, T, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3302</th>\n",
       "      <td>CpxQ(ncRNA0205)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>4271116</td>\n",
       "      <td>4271175</td>\n",
       "      <td>4.130000e-26</td>\n",
       "      <td>111.0</td>\n",
       "      <td>SP</td>\n",
       "      <td>(G, T, T, T, T, C, C, T, T, G, C, C, A, T, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>RprA(ncRNA0058)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>51</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>1444873</td>\n",
       "      <td>1444823</td>\n",
       "      <td>3.250000e-21</td>\n",
       "      <td>95.3</td>\n",
       "      <td>SP</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>CpxQ(ncRNA0205)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>4271117</td>\n",
       "      <td>4271176</td>\n",
       "      <td>4.130000e-26</td>\n",
       "      <td>111.0</td>\n",
       "      <td>SP</td>\n",
       "      <td>(T, T, T, T, C, C, T, T, G, C, C, A, T, A, G, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3305</th>\n",
       "      <td>RprA(ncRNA0058)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>1444894</td>\n",
       "      <td>1444821</td>\n",
       "      <td>9.150000e-34</td>\n",
       "      <td>137.0</td>\n",
       "      <td>SP</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3306</th>\n",
       "      <td>SdsR(ncRNA0070)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>75</td>\n",
       "      <td>1968124</td>\n",
       "      <td>1968050</td>\n",
       "      <td>2.590000e-34</td>\n",
       "      <td>139.0</td>\n",
       "      <td>SP</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6036 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Query_ID   Subject_ID  PIdentity  Alignment_Length  \\\n",
       "0         ArcZ(ncRNA0002)  NC_003197.2      100.0                58   \n",
       "1     STnc2110(ncRNA0286)  NC_003197.2      100.0                49   \n",
       "2         CpxQ(ncRNA0205)  NC_003197.2      100.0                61   \n",
       "3     STnc2000(ncRNA0278)  NC_003197.2      100.0                43   \n",
       "4         ArcZ(ncRNA0002)  NC_003197.2      100.0                58   \n",
       "...                   ...          ...        ...               ...   \n",
       "3302      CpxQ(ncRNA0205)  NC_003197.2      100.0                60   \n",
       "3303      RprA(ncRNA0058)  NC_003197.2      100.0                51   \n",
       "3304      CpxQ(ncRNA0205)  NC_003197.2      100.0                60   \n",
       "3305      RprA(ncRNA0058)  NC_003197.2      100.0                74   \n",
       "3306      SdsR(ncRNA0070)  NC_003197.2      100.0                75   \n",
       "\n",
       "      Mismatches  Gap_Openings  Query_Start  Query_End  Subject_Start  \\\n",
       "0              0             0            1         58        3490451   \n",
       "1              0             0            1         49        4009394   \n",
       "2              0             0            1         61        4271116   \n",
       "3              0             0            1         43        1248179   \n",
       "4              0             0            1         58        3490451   \n",
       "...          ...           ...          ...        ...            ...   \n",
       "3302           0             0            1         60        4271116   \n",
       "3303           0             0            1         51        1444873   \n",
       "3304           0             0            1         60        4271117   \n",
       "3305           0             0            1         74        1444894   \n",
       "3306           0             0            1         75        1968124   \n",
       "\n",
       "      Subject_End       E_value  Bit_Score scenario  \\\n",
       "0         3490508  5.090000e-25      108.0       EP   \n",
       "1         4009346  3.940000e-20       91.6       EP   \n",
       "2         4271176  1.180000e-26      113.0       EP   \n",
       "3         1248137  7.110000e-17       80.5       EP   \n",
       "4         3490508  5.090000e-25      108.0       EP   \n",
       "...           ...           ...        ...      ...   \n",
       "3302      4271175  4.130000e-26      111.0       SP   \n",
       "3303      1444823  3.250000e-21       95.3       SP   \n",
       "3304      4271176  4.130000e-26      111.0       SP   \n",
       "3305      1444821  9.150000e-34      137.0       SP   \n",
       "3306      1968050  2.590000e-34      139.0       SP   \n",
       "\n",
       "                                            Subject_Seq  \n",
       "0     (A, T, T, T, C, C, C, T, G, G, T, G, T, T, G, ...  \n",
       "1                                                    ()  \n",
       "2     (G, T, T, T, T, C, C, T, T, G, C, C, A, T, A, ...  \n",
       "3                                                    ()  \n",
       "4     (A, T, T, T, C, C, C, T, G, G, T, G, T, T, G, ...  \n",
       "...                                                 ...  \n",
       "3302  (G, T, T, T, T, C, C, T, T, G, C, C, A, T, A, ...  \n",
       "3303                                                 ()  \n",
       "3304  (T, T, T, T, C, C, T, T, G, C, C, A, T, A, G, ...  \n",
       "3305                                                 ()  \n",
       "3306                                                 ()  \n",
       "\n",
       "[6036 rows x 14 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "salmonella_genome_sz = len(salmonella_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_mismatch_alignments = alignments_df[alignments_df[\"PIdentity\"] == 100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA1_seq_matches = no_mismatch_alignments[\"Subject_Seq\"].isin(sup5_data_df[\"RNA1 seq\"])\n",
    "RNA2_seq_matches = no_mismatch_alignments[\"Subject_Seq\"].isin(sup5_data_df[\"RNA2 seq\"])\n",
    "\n",
    "not_matched = no_mismatch_alignments[~(RNA1_seq_matches | RNA2_seq_matches)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Subject_Seq\n",
       "0      2312\n",
       "72       15\n",
       "84       13\n",
       "52       12\n",
       "57       11\n",
       "79        9\n",
       "54        9\n",
       "55        8\n",
       "81        6\n",
       "85        5\n",
       "56        5\n",
       "87        4\n",
       "46        4\n",
       "86        4\n",
       "48        3\n",
       "28        3\n",
       "64        2\n",
       "68        2\n",
       "53        2\n",
       "74        2\n",
       "47        2\n",
       "80        2\n",
       "73        2\n",
       "100       2\n",
       "99        2\n",
       "59        2\n",
       "71        2\n",
       "66        1\n",
       "89        1\n",
       "37        1\n",
       "45        1\n",
       "36        1\n",
       "29        1\n",
       "65        1\n",
       "42        1\n",
       "41        1\n",
       "69        1\n",
       "92        1\n",
       "90        1\n",
       "75        1\n",
       "31        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_matched.Subject_Seq.map(len).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
