{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from Bio import AlignIO\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import re\n",
    "\n",
    "from src.sup_data_to_fasta import load_xlsx\n",
    "\n",
    "sup5_file_path = Path(\"data/Liu_sup5_data.xlsx\")\n",
    "alignments_files_paths = [\n",
    "    Path(\"output/EP-OD0_alignments_results.tsv\"),\n",
    "    Path(\"output/ESP-OD2_alignments_results.tsv\"),\n",
    "    Path(\"output/SP-OD2_alignments_results.tsv\")\n",
    "    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Análise Exploratória dos Dados Suplementares 5\n",
    "\n",
    "## Sobre\n",
    "\n",
    "Este notebook contém uma análise exploratória dos dados fornecidos por Liu, et al. No [suplemento 5](https://www.nature.com/articles/s41467-023-43632-1#additional-information\n",
    ") com o objetivo de averiguar os tratamentos executados no artigo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_scenario_column(df_dict):\n",
    "    return {k:df.assign(scenario=k) for k,df in df_dict.items()}\n",
    "\n",
    "def extract_scenario_from_file_path(string_arg):\n",
    "    return re.search(r\"(.*)-(.*)\", string_arg).group(1)\n",
    "\n",
    "def concat_df_dict(df_dict):\n",
    "    return pd.concat(df_dict.values())\n",
    "\n",
    "alignments_df_list = [pd.read_csv(file_path, sep='\\t') for file_path in alignments_files_paths]\n",
    "alignments_df_dict = {extract_scenario_from_file_path(k.name):v for k,v in zip(alignments_files_paths, aligments_df_list)}\n",
    "alignments_df_dict = add_scenario_column(alignments_df_dict)\n",
    "alignments_df = concat_df_dict(alignments_df_dict)\n",
    "\n",
    "sup5_data_df_dict = load_xlsx(sup5_file_path)\n",
    "sup5_data_df_dict = {extract_scenario_from_file_path(k):v for k,v in sup5_data_df_dict.items()}\n",
    "sup5_data_df_dict = add_scenario_column(sup5_data_df_dict)\n",
    "sup5_data_df = concat_df_dict(sup5_data_df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query_ID</th>\n",
       "      <th>Subject_ID</th>\n",
       "      <th>PIdentity</th>\n",
       "      <th>Alignment_Length</th>\n",
       "      <th>Mismatches</th>\n",
       "      <th>Gap_Openings</th>\n",
       "      <th>Query_Start</th>\n",
       "      <th>Query_End</th>\n",
       "      <th>Subject_Start</th>\n",
       "      <th>Subject_End</th>\n",
       "      <th>E_value</th>\n",
       "      <th>Bit_Score</th>\n",
       "      <th>scenario</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>3490451</td>\n",
       "      <td>3490508</td>\n",
       "      <td>5.090000e-25</td>\n",
       "      <td>108.0</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STnc2110(ncRNA0286)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>4009394</td>\n",
       "      <td>4009346</td>\n",
       "      <td>3.940000e-20</td>\n",
       "      <td>91.6</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CpxQ(ncRNA0205)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>4271116</td>\n",
       "      <td>4271176</td>\n",
       "      <td>1.180000e-26</td>\n",
       "      <td>113.0</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STnc2000(ncRNA0278)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>43</td>\n",
       "      <td>1248179</td>\n",
       "      <td>1248137</td>\n",
       "      <td>7.110000e-17</td>\n",
       "      <td>80.5</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ArcZ(ncRNA0002)</td>\n",
       "      <td>NC_003197.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>3490451</td>\n",
       "      <td>3490508</td>\n",
       "      <td>5.090000e-25</td>\n",
       "      <td>108.0</td>\n",
       "      <td>EP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Query_ID   Subject_ID  PIdentity  Alignment_Length  Mismatches  \\\n",
       "0      ArcZ(ncRNA0002)  NC_003197.2      100.0                58           0   \n",
       "1  STnc2110(ncRNA0286)  NC_003197.2      100.0                49           0   \n",
       "2      CpxQ(ncRNA0205)  NC_003197.2      100.0                61           0   \n",
       "3  STnc2000(ncRNA0278)  NC_003197.2      100.0                43           0   \n",
       "4      ArcZ(ncRNA0002)  NC_003197.2      100.0                58           0   \n",
       "\n",
       "   Gap_Openings  Query_Start  Query_End  Subject_Start  Subject_End  \\\n",
       "0             0            1         58        3490451      3490508   \n",
       "1             0            1         49        4009394      4009346   \n",
       "2             0            1         61        4271116      4271176   \n",
       "3             0            1         43        1248179      1248137   \n",
       "4             0            1         58        3490451      3490508   \n",
       "\n",
       "        E_value  Bit_Score scenario  \n",
       "0  5.090000e-25      108.0       EP  \n",
       "1  3.940000e-20       91.6       EP  \n",
       "2  1.180000e-26      113.0       EP  \n",
       "3  7.110000e-17       80.5       EP  \n",
       "4  5.090000e-25      108.0       EP  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def should_be_only_one_genome(df):\n",
    "    return df.Subject_ID.nunique() == 1\n",
    "\n",
    "def query_start_should_always_be_less_than_query_end(df):\n",
    "    return (df.Query_Start <= df.Query_End).all()\n",
    "\n",
    "# validationg\n",
    "assert should_be_only_one_genome(alignments_df), \"o banco de dados deveria ter apenas um genoma de referência de Salmonella\" # all alignments are done to same Salmonella enterica genome\n",
    "assert query_start_should_always_be_less_than_query_end(alignments_df), \"O ínicio da query deve ser menor ou igual ao fim da query\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# queries não alinhadas: 0\n"
     ]
    }
   ],
   "source": [
    "def not_aligned(col):\n",
    "    return col[~col.isin(alignments_df.Query_ID)]\n",
    "\n",
    "RNA1_not_aligned = not_aligned(sup5_data_df[\"RNA1 name\"])\n",
    "RNA2_not_aligned = not_aligned(sup5_data_df[\"RNA2 name\"])\n",
    "\n",
    "filtered_alignments_df = alignments_df[alignments_df['Query_ID'].isin(RNA1_not_aligned) & alignments_df['Query_ID'].isin(RNA2_not_aligned)]\n",
    "print(f\"# queries não alinhadas: {len(filtered_alignments_df)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
